### PR TITLE
Make multiple threads creating their own sky.Dag() thread-safe.

### PR DESCRIPTION
**Pull request by concretevitamin**
Thanks to the report by @Michaelvll, who saw that test_multiecho is failing:
```
[?25hTraceback (most recent call last):
  File "examples/multi_echo.py", line 45, in <module>
    run(cluster)
  File "examples/multi_echo.py", line 37, in run
    list(p.imap(_exec, range(32)))
  File "/Users/zhwu/miniconda3/envs/sky-dev/lib/python3.8/multiprocessing/pool.py", line 868, in next
    raise value
  File "/Users/zhwu/miniconda3/envs/sky-dev/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "examples/multi_echo.py", line 34, in _exec
    sky.exec(dag, cluster_name=cluster, detach_run=True)
  File "/Users/zhwu/Library/CloudStorage/OneDrive-Personal/AResource/PhD/Research/sky-computing/code/sky-experiment-dev/sky/utils/common_utils.py", line 129, in _record
    return f(*args, **kwargs)
  File "/Users/zhwu/Library/CloudStorage/OneDrive-Personal/AResource/PhD/Research/sky-computing/code/sky-experiment-dev/sky/execution.py", line 249, in exec
    _execute(dag=dag,
  File "/Users/zhwu/Library/CloudStorage/OneDrive-Personal/AResource/PhD/Research/sky-computing/code/sky-experiment-dev/sky/execution.py", line 93, in _execute
    assert len(dag) == 1, f'We support 1 task for now. {dag}'
AssertionError: We support 1 task for now. DAG:
[]
```
Reading the code, 

https://github.com/skypilot-org/skypilot/blob/a04d14922de8dfd0ec5351b64a65eb61c6290b36/examples/multi_echo.py#L23-L33

this is likely because `sky.Dag() -> DagContext` wasn't thread-safe. This PR makes it so.

Tested:
- `examples/multi_echo.py` multiple times [Note: I couldn't reproduce the above bug]

### CODE REVIEW

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick fix @concretevitamin! Seems L43 and L44 are not used. Should we remove them?

It seems to me the error shown above may be caused by these lines, where the current_dag is not the one the task is in, as the `DagContext._current_dag` may be changed by another thread, before the task is trying to get the current dag.
https://github.com/skypilot-org/skypilot/blob/b8abd835f233bbcbcc65b6648b8d4ddc49e6f77a/sky/task.py#L146-L148

**Comment by concretevitamin** [comment]
@Michaelvll That's a great catch! PTAL.

File: sky/dag.py [hunk 1]
```diff
@@ -1,18 +1,16 @@
...
 import pprint
+import threading
 
 
-class DagContext:
-    """A global stack of Dags.
-
-    Currently, we only use one sky.Dag.
-    """
+class DagContext(threading.local):
```
Comment (Michaelvll) [inline sky/dag.py:6]: Just curious, will this work if it is not instantiated? Should we create a global variable in this file `dag_context = DAGContext()` and make the classmethod to be normal class member methods?

Comment (concretevitamin) [reply sky/dag.py]: Great catch! Pushed the changes, PTAL.

Re-ran smoke tests, all passed except:
- test_spot_storage: known access issue
- test_use_spot, test_spot: same issues as #1026,  keypairs in us-east-2

**Review by Michaelvll — APPROVED** [review]
LGTM @concretevitamin ! Thank you for fixing this!

**Pull request merged**